[windows] Add kitchen tests for cws

derekwbrown opened the pull request:
### What does this PR do?

Add ktichen test for basic install of CWS

### Motivation


### Additional Notes


### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

derekwbrown committed:
[windows] Add kitchen tests for cws

derekwbrown committed:
revert this change prior to merging; testing build with cws excluded

derekwbrown committed:
Revert "revert this change prior to merging; testing build with cws excluded"

This reverts commit b0c3c7f4c56f24fd723ca244059849a18b103ef7.

richl-dd commented on test/kitchen/test/integration/win-agent-with-cws-option/rspec_datadog/win-agent-with-cws-option_spec.rb:
```diff
@@ -0,0 +1,131 @@
+            expect(is_windows_service_installed("ddprocmon")).to be_truthy
+
+            program_files = safe_program_files
+            expect(File).to exist("#{program_files}\\DataDog\\Datadog Agent\\bin\\agent\\ddprocmon.cat")
+            expect(File).to exist("#{program_files}\\DataDog\\Datadog Agent\\bin\\agent\\ddprocmon.sys")
+            expect(File).to exist("#{program_files}\\DataDog\\Datadog Agent\\bin\\agent\\ddprocmon.inf")
```
Are these paths correct? If so, why are the `ddprocmon.*` files in `bin\agent` and not in `bin\agent\driver` with the rest of the drivers?

derekwbrown replied:
they're currently correct, as the procmon merge module isn't putting them in the right place.
Need to fix the procmon merge module, and then would (in a separate pr) update paths to match

richl-dd commented on test/kitchen/test/integration/win-agent-with-cws-option/rspec_datadog/win-agent-with-cws-option_spec.rb:
```diff
@@ -0,0 +1,131 @@
+            expect(is_windows_service_installed("ddprocmon")).to be_falsey
+
+            program_files = safe_program_files
+            expect(File).not_to exist("#{program_files}\\DataDog\\Datadog Agent\\bin\\agent\\ddprocmon.cat")
+            expect(File).not_to exist("#{program_files}\\DataDog\\Datadog Agent\\bin\\agent\\ddprocmon.sys")
+            expect(File).not_to exist("#{program_files}\\DataDog\\Datadog Agent\\bin\\agent\\ddprocmon.inf")
```
Same file path question as above.

richl-dd commented on test/kitchen/test/integration/win-agent-with-cws-option/rspec_datadog/win-agent-with-cws-option_spec.rb:
```diff
@@ -0,0 +1,131 @@
+            expect(is_service_running?("datadog-security-agent")).to be_falsey
+            #restart "datadog-agent"
+            stop "datadog-agent"
+            sleep 10
```
Can we replace the `sleep` with a periodic check as to if the the service stopped and then only proceed if it stopped within an acceptable time period? Arbitrary sleeps can contribute to flaky tests, especially when it comes to Windows service shutdown and startup.

I think this is important because these are new Windows Services and there very well could be a bug in the start/stop/restart behavior.

For example (pseudo code) could be have something like:
```
def wait_for(action, checkIntervalInSeconds, maxTimeoutInSeconds)

if(!wait_for(service_stopped("datadog-agent", 1, 10)):
  fail the test, the action didn't happen within the specified maxTimeout period
```

This I think will bound the time we expect the services to startup / shutdown and not proceed with the test if/when that action failed.

I recognize that the `maxTimeout` is similar to an arbitrary sleep, but I think the benefit here is that we won't proceed if the timeout is exceeded.

richl-dd replied:
With the above in mind, if the subsequent call to `start` would fail / throw an error if the service was still "started", then we might not need the above. But if it silently continues, then the `expect` calls below could find that the service is running, but it is running because it never stopped instead of knowing that it stopped, was started, and is now running again.

derekwbrown replied:
so I'm not sure if I can do that in ruby.  I suppose I could. But as you point out
- waitfor(arbitrary time) is functionally the same as (sleep for) arbitrary time
- and either way, if the service isn't stopped at the end of the arbitrary time period, the test will fail and not proceed.

So I'm missing the improvement.

richl-dd replied:
The improvement is that we can control the max timeout we're willing to accept on shutdown and then if that is exceeded, we know that it was because it happened as a direct result of not occurring soon enough.

> and either way, if the service isn't stopped at the end of the arbitrary time period, the test will fail and not proceed.

Is that the behavior right now? As written the test is:
```
stop
sleep
start
sleep
check_is_still_running
```
When we check that it is still running, how do we know that it is running after a successful stop / start vs. the stop failing and it ran the entire time?

derekwbrown replied:
The stop() and start() functions achieve the desired state or fail.

richl-dd replied:
To close the thread, the sleeps that were unneeded were removed and the remaining `sleep 30` gives enough time for agent to get into steady state before shutting it down again.

richl-dd commented on test/kitchen/test/integration/win-agent-with-cws-option/rspec_datadog/win-agent-with-cws-option_spec.rb:
```diff
@@ -0,0 +1,131 @@
+            end
+            f = File.read(sa_conf_path)
+            confYaml = YAML.load(f)
+            if !confYaml
```
Why do we proceed with the remainder of the test if there isn't a valid `confYaml`? If we expect CWS to be installed, then the `.yaml` files should have been written already and if they aren't, then shouldn't the test fail here? Otherwise it seem we're artificially allowing the test to continue after we realize the files aren't there.

Big caveat in that I'm not completely well versed with kitchen - on my first read though, it seems like this should be a fail case.

derekwbrown replied:
I suppose we could make it a fail case.  But we don't have "warnings" or "fail and continue".
So since we can continue in this case and see that the service starts...
But, i see your point and happy to change.

richl-dd replied:
Thanks, yeah, I think that we should make this (and any other checks that are required before we'd start the service) a fail case. We could separate it out into different functions like:

```
if cws_expected:
    if !verifyCWSFilesWritten(): // check they exist at the expected paths
        fail("expected files not found)

    if !verifyCWSFileContent(): // check they have the right key/values
       fail("malformed default files")

    enableCWSInFiles() // change whatever flags we need to enabled: true

    // continue to check that the services started up  
```

derekwbrown replied:
I put in the checks for existence.
However, I left the existing error checking code in place.  Our default files for both of these are entirely commented out, which means that they're effectively an empty file.  So even if the file is present as expected, we have to build the correct config entrires.

richl-dd replied:
Ah that's annoying - are there tests somewhere else validate the default config is correct before everything is commented out? I mainly want to make sure that the default (albeit commented out) config is correct so that when customers comment out the key/values, the commented out key/values are correct.

richl-dd commented on test/kitchen/test/integration/win-agent-with-cws-option/rspec_datadog/win-agent-with-cws-option_spec.rb:
```diff
@@ -0,0 +1,131 @@
+            if !confYaml
+                confYaml = {}
+            end
+            if !confYaml.key("runtime_security_config")
+                confYaml["runtime_security_config"] = {}
+            end
+            confYaml["runtime_security_config"]["enabled"] = true
+            File.write(sa_conf_path, confYaml.to_yaml)
+        
+            spf = File.read(sp_conf_path)
+            spconfYaml = YAML.load(spf)
+            if !spconfYaml
+                spconfYaml = {}
+            end
+            if !spconfYaml.key("runtime_security_config")
+                spconfYaml["runtime_security_config"] = {}
+            end
+            spconfYaml["runtime_security_config"]["enabled"] = true
+            File.write(sp_conf_path, spconfYaml.to_yaml)
```
Same sentiment as comment above (https://github.com/DataDog/datadog-agent/pull/21022/files#r1402432742).

If the default files not written to disk and/or don't contain the keys we expect, then I think the test should fail here because that implies a bad install.

Why do we take corrective actions in the case that the files aren't what we expect them to be and then continue the installation / checks?

derekwbrown replied:
The code as written is just to prevent ruby from crashing.  We could accomplish that, and do what you want, by early failing.

richl-dd replied:
Okay, then as discussed above, let's early fail 👍

richl-dd commented on test/kitchen/test/integration/win-agent-with-cws-option/rspec_datadog/win-agent-with-cws-option_spec.rb:
```diff
@@ -0,0 +1,131 @@
+    it_behaves_like 'an installed Agent'
+    it_behaves_like 'a running Agent with no errors'
+    it_behaves_like 'a Windows Agent with CWS driver installed'
+    it_behaves_like 'a Windows Agent with CWS driver disabled'
```
Why is this included in this list? the surrounding `it_behaves_like` seem to verify that the cws components are good to go but this one implies that there are still parts that aren't supposed to be running.

derekwbrown replied:
which one?

richl-dd replied:
I _did_ highlight it but Github didn't like it, I guess.

The `a Windows Agent with CWS driver disabled` one.

derekwbrown replied:
so, the logic based on the way we do drivers
- fresh install (no config changes)
  - agent should be running `it behaves like a running agent no errors
  - driver should be installed `it behaves like a windows agent with cws driver installed`
  - driver should be disabled on initial install; it's set to disabled and the system probe moves it from disabled to manual start on demand `it behaves liek a windows agent with cws driver disabled`
  
  So the test at this point is validating that the result of an install w/o changing configs has all the pieces in the expected state.

richl-dd reviewed (commented):
This PR covers install and start up from what I can tell - can we add additional tests for shutdown, uninstall, upgrade?

derekwbrown commented:
> This PR covers install and start up from what I can tell - can we add additional tests for shutdown, uninstall, upgrade?

It already tests that security agent & system probe can stop (which was a bug at one point).
There are already separate kitchen upgrade tests, and uninstall tests.  what did you have in mind?

derekwbrown committed:
add test for config file existence

richl-dd commented:
> > This PR covers install and start up from what I can tell - can we add additional tests for shutdown, uninstall, upgrade?
> 
> It already tests that security agent & system probe can stop (which was a bug at one point). There are already separate kitchen upgrade tests, and uninstall tests. what did you have in mind?

Do the separate upgrade and uninstall tests verify that the new files (drivers, configs) and services (security agent) are removed on uninstall? Do they verify that on upgrade, the new files are upgraded?

derekwbrown committed:
remove some of the sleeps

derekwbrown commented:
> > > This PR covers install and start up from what I can tell - can we add additional tests for shutdown, uninstall, upgrade?
> > 
> > 
> > It already tests that security agent & system probe can stop (which was a bug at one point). There are already separate kitchen upgrade tests, and uninstall tests. what did you have in mind?
> 
> Do the separate upgrade and uninstall tests verify that the new files (drivers, configs) and services (security agent) are removed on uninstall? Do they verify that on upgrade, the new files are upgraded?

the uninstall test will verify that all the files are removed on uninstall.

The upgrade tests don't know anything about upgrade.  However, since the driver version is not tied to the agent version, then upgrade from one version to the next may not upgrade the driver.

richl-dd commented on test/kitchen/test/integration/win-agent-with-cws-option/rspec_datadog/win-agent-with-cws-option_spec.rb:
```diff
@@ -0,0 +1,134 @@
+        
+            expect(is_service_running?("datadog-system-probe")).to be_falsey
+            expect(is_service_running?("datadog-security-agent")).to be_falsey
+            #restart "datadog-agent"
```
Was this commented out by mistake? Implying we wanted to test the restart behavior and then proceed to stop/start? Or is it left behind commented out code that can be removed?

derekwbrown replied:
it was copy paste error I removed.

derekwbrown committed:
remove copy paste error

derekwbrown committed:
add check for removing all services

derekwbrown committed:
redo CWS tests using different `if cws expected` method
Allows breaking up into more readable `it` blocks for different segments of
the test.

derekwbrown committed:
further condense & make readable by using existing helper
functions to find & change config files.
separate out existence check of config file into separate `it` block
from the modify/enable cws block

pducolin commented:
What is preventing implementing this new test in the new e2e test framework ? Is it lack of required features, such as local VM support ? Or is it lack of documentation and / or familiarity with the new framework ? I can help if it's the latter, I can help with the implementation and improving docs - @KevinFairise2  recently added [installer tests](https://github.com/DataDog/datadog-agent/blob/main/test/new-e2e/tests/agent-platform/install-script/install_script_test.go), we can work together to adapt them for supporting windows.

If it's the first, support is coming soon, ETA end of the quarter, possibly sooner. Does it make sense to initially implement this tests on remote VMs only ?

derekwbrown committed:
fix error parsing whether cws is to be included

pducolin reviewed (approved):
As discussed in private, plans are to eventually migrate to new e2e framework likely across agent teams. Currently the lack of support for running tests on a local VM and lack of familiarity and time to familiarise with the framework, plus urgency to provide coverage for cws, makes merging this PR before new e2e adoption a priority.
